### ISSUE TITLE
Add mypy static type hinting

### ISSUE THREAD

**Issue by pganssle** [dateutil/dateutil#383]
To facilitate downstream users who want to use static type hinting in their codebase, it seems like it's a good idea to use the `mypy`-style static type hinting in the `dateutil` codebase.  Some of this work has already been done in the [`typeshed` library](https://github.com/python/typeshed/tree/master/third_party/3/dateutil) ([2](https://github.com/python/typeshed/tree/master/third_party/2/dateutil)), assuming the license there is compatible with 3-clause BSD we may be able to use that.

To Do:
[ ] Add strict mypy checking to the CI.
[ ] Determine what might need to be done in the 2 vs. 3 case (I think they may be slightly different)

Modules to cover:
[ ] easter
[ ] parser
[ ] relativedelta
[ ] rrule
[ ] tz
[ ] zoneinfo

I am not sure if we want to do this entirely inline or with stubs - anyone who knows mypy maybe can chime in on the benefits of each. I would think the ideal would be to use the inline comment-style hinting, but I'm not sure how that works with 2/3 polyglot code.

cc @theghosthucodes who has volunteered to help out with this effort.

**Comment by TheGhostHuCodes** [comment]
Yeah, it looks like adding type hints to the project repo is the preferred way to go (https://github.com/python/typeshed/blob/master/CONTRIBUTING.md#adding-a-new-library) according to `typeshed`.

I've worked mostly with python 2.7 compatible "comment style" type hints within the `*.py` file and that seams to me like the easiest way to display and maintain the type hints in 2/3 polyglot:

 - You don't have to restate the function signature in a separate `*.pyi` file
 - You don't need an extra file open when you're working with code

**Comment by pganssle** [comment]
Yeah, I like the idea of doing it inline. I'm guessing that the biggest 2/3 problem would be things that explicitly will have a return value of `six.text_type` or the like, since `[str,unicode]` is wrong for those. I'm guessing we can actually declare them as `[six.text_type]`?

**Comment by TheGhostHuCodes** [comment]
Yeah, there's a type that's available in `typing` called `Text` that we can use for this. It is available as part of the standard library in python 3.5 I think, but will add a dependency for python 2.7 to the 3rd party `typing` library.

**Comment by wagner-certat** [comment]
> Yeah, there's a type that's available in typing called Text that we can use for this. It is available as part of the standard library in python 3.5 I think, but will add a dependency for python 2.7 to the 3rd party typing library.

And also for 3.x < 3.5

**Comment by pganssle** [comment]
Hmmm... I'm much more comfortable adding a required dependency for 2.7 than for 3.4.

I'm wondering if, for compatibility reasons, we should just go ahead and use `.pyi` files - at least until we drop support for 2.x; that way we're not adding a dependency at all for people *not* using mypy. Other than the duplication of the function signatures (which we can probably catch as part of CI), I don't think there's a major downside to that.

**Comment by TheGhostHuCodes** [comment]
@wagner-certat, good point! There might be some complex conditional way of importing `typing` within the `*.py` file as well, but that's probably more brittle and error prone than just using the `*pyi` file and adding types to the function signatures. I'll try and get `mypy` setup first in the CI, but then maybe we should just talk to the [typeshed](https://github.com/python/typeshed) people and see if we can just wholesale move their `*.pyi` files over. Typeshed license is Apache-2.

**Comment by gaborbernat** [comment]
I think this would use the newly added https://github.com/python/mypy/issues/5028 - this would allow both typeshed and dateutil to benefit from things.

**Comment by anthrotype** [comment]
FYI I'm trying to run [monkeytype](https://github.com/Instagram/MonkeyType) on dateutils, just for fun.
I must stay I've never used typing annotation myself yet, but I'm eager to learn, especially after reading @gaborbernat's nice write up on https://www.bernat.tech/the-state-of-type-hints-in-python/

**Comment by Zac-HD** [comment]
We've been looking at this for a while in HypothesisWorks/hypothesis#200.  The short version is that comments are currently the best way to distribute type hints to users (see PEP561), because annotations don't work on Python 2 and stubs are difficult to keep in sync with the implementation code (WIP, but a long way from stable).

All you need to do is gate the typing (or circular) imports behind a literal `if False:`, and they'll be picked up by Mypy but have no effect at runtime.

**Comment by jpurviance** [comment]
Is the addition of type hinting to dateutil strictly for the use of those that want to run a static type checker (like `mypy` or facebook's `pyre` project) against the package? That's great for those who would like to use the project. If annotations are being added to the project, it might make sense to annotate more than the public methods and functions. `Mypy` can tell us a lot about the way the codebase works and unexpected errors.

**Comment by jpurviance** [comment]
See my first pass at annotations for `utils.py`: https://github.com/dateutil/dateutil/pull/768
It's always best to start adding annotations to modules that depend on very little and are depended on by many others.

**Comment by udalrich** [comment]
Is there any progress likely to happen on this?  I'm still not able to use Mypy and dateutil in the same project

**Comment by StefanBrand** [comment]
Maybe this helps: https://pypi.org/project/types-python-dateutil/

**Comment by kennylajara** [comment]
> Maybe this helps: https://pypi.org/project/types-python-dateutil/
Nope. Is not helping. Looks like a bug.

**Comment by Coinhexa** [comment]
Problem still exists
```

mypy.....................................................................Failed
- hook id: mypy
- duration: 3.13s
- exit code: 1

src/news/utils.py:7:1: error: Library stubs not installed for "dateutil" (or
incompatible with Python 3.10)  [import]
    from dateutil import parser
    ^
src/news/utils.py:7:1: note: Hint: "python3 -m pip install types-python-dateutil"
src/news/utils.py:7:1: note: (or run "mypy --install-types" to install all missing stub packages)
src/news/utils.py:7:1: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
src/news/utils.py: note: In function "build_pubdate":
src/news/utils.py:51:5: error: Returning Any from function declared to return
"datetime"  [no-any-return]
        return parser.parse(pubdate)
        ^
src/news/parser.py:4:1: error: Library stubs not installed for
"dateutil.parser" (or incompatible with Python 3.10)  [import]
    from dateutil.parser import ParserError
    ^
Found 3 errors in 2 files (checked 6 source files)
```
The funny thing is I have an ignore_missing_imports=true set inside pyproject.toml I have no idea why mypy from pre-commit wont read this setting

**Comment by Dor-bl** [comment]
Same issue for me as well.
```
(venv) E:\Git\python-client>mypy --python-executable venv\Scripts\python --install-types
Installing missing stub packages:
venv\Scripts\python -m pip install types-python-dateutil

Install? [yN] y

Requirement already satisfied: types-python-dateutil in e:\git\python-client\venv\lib\site-packages (2.8.19.13)

(venv) E:\Git\python-client>git status
On branch selendroid_Cleanup
Changes to be committed:
  (use "git restore --staged <file>..." to unstage)
        modified:   test/functional/android/applications_tests.py


(venv) E:\Git\python-client>git commit m "test: replace usage of selendroid app from 'test_install_app' in applications_tests.py"
error: pathspec 'm' did not match any file(s) known to git
error: pathspec 'test: replace usage of selendroid app from 'test_install_app' in applications_tests.py' did not match any file(s) known to git

(venv) E:\Git\python-client>git commit -m "test: replace usage of selendroid app from 'test_install_app' in applications_tests.py"
isort....................................................................Passed
mypy.....................................................................Failed
- hook id: mypy
- exit code: 1

test\functional\android\device_time_tests.py:16: error: Library stubs not installed for "dateutil.parser"  [import]
test\functional\android\device_time_tests.py:16: note: Hint: "python3 -m pip install types-python-dateutil"
test\functional\android\device_time_tests.py:16: note: (or run "mypy --install-types" to install all missing stub packages)
test\functional\android\device_time_tests.py:16: note: See https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports
Found 1 error in 1 file (checked 361 source files)
```
